### ISSUE TITLE
Runtime error during shallow fusion

### ISSUE THREAD

**Issue by bsshruthi22** [k2-fsa/icefall#1235]
Trained a pruned transducer stateless 7 model for English & Hindi dataset.Also trained a RNN language model on domain data.Trying to perform shallow fusion.
After running the below command ,I get error as follows:

Decode command:

python pruned_transducer_stateless7/decode.py \
    --epoch 30 \
    --avg 15 \
    --exp-dir ./pruned_transducer_stateless7/exp \
    --max-duration 600 \
    --decoding-method modified_beam_search_lm_shallow_fusion \
    --beam-size 4 \
    --lm-type rnn \
    --lm-scale 0.3 \
    --use-shallow-fusion True \
    --lm-exp-dir ./pruned_transducer_stateless7/rnn_lm/exp 

RNN Training command:
    ./rnn_lm/train.py \
        --start-epoch 0 \
        --world-size 2 \
        --num-epochs 1 \
        --use-fp16 0 \
        --tie-weights 0 \
        --embedding-dim 800 \
        --hidden-dim 200 \
        --num-layers 2 \
        --batch-size 400
Error:
raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for RnnLmModel:
size mismatch for input_embedding.weight: copying a param with
shape torch.Size([500, 800]) from checkpoint, the shape in current model is torch.Size([500, 2048]).

The rnn model dimension is taken as 2048,even though its trained with dim 800.It is taking internal LM 
dimension.
In decode.py script,main function has this update line, params.update(vars(args)) ,
 where the internal parameters are getting updated.
Please let me know how to resolve this issue.

**Comment by csukuangfj** [comment]
There is an argument, --tie-weights. 

please check that decode.py uses the same argument value as train.py.

(Note: The default argument values may not be the same. You need to check that.)

**Comment by bsshruthi22** [comment]
Not able to find that parameter in both train.py and decode.py
icefall/egs/librispeech/ASR/pruned_transducer_stateless7/train.py
icefall/egs/librispeech/ASR/pruned_transducer_stateless7/decode.py

**Comment by bsshruthi22** [comment]
@csukuangfj  Can you please let me know in which file I can find the parameter --tie-weights. Thanks.

**Comment by csukuangfj** [comment]
```
Decode command:

python pruned_transducer_stateless7/decode.py
--epoch 30
--avg 15
--exp-dir ./pruned_transducer_stateless7/exp
--max-duration 600
--decoding-method modified_beam_search_lm_shallow_fusion
--beam-size 4
--lm-type rnn
--lm-scale 0.3
--use-shallow-fusion True
--lm-exp-dir ./pruned_transducer_stateless7/rnn_lm/exp
```

`python pruned_transducer_stateless7/decode.py` does not support the argument `--lm-exp-dir`.
Could you tell us how did you get the error?

**Comment by bsshruthi22** [comment]
@csukuangfj    The argument --lm-exp-dir was in documentation. When I run without this argument,I get error ,as it taken rnn exp dir as None.below is the error:
python pruned_transducer_stateless7/decode.py --epoch 30 \
> --avg 15 \
> --exp-dir ./pruned_transducer_stateless7/exp \
> --max-duration 600 \
> --decoding-method modified_beam_search_lm_shallow_fusion \
> --beam-size 4 \
> --lm-type rnn \
> --lm-scale 0.3 \
> --use-shallow-fusion True
2023-09-04 15:45:08,866 INFO [decode.py:777] Decoding started
2023-09-04 15:45:08,867 INFO [decode.py:783] Device: cuda:0
2023-09-04 15:45:08,868 INFO [decode.py:793] {'frame_shift_ms': 10.0, 'allowed_excess_duration_ratio': 0.1, 'best_train_loss': inf, 'best_valid_loss': inf, 'best_train_epoch': -1, 'best_valid_epoch': -1, 'batch_idx_train': 0, 'log_interval': 50, 'reset_interval': 200, 'valid_interval': 3000, 'feature_dim': 80, 'subsampling_factor': 4, 'warm_step': 2000, 'env_info': {'k2-version': '1.23.4', 'k2-build-type': 'Release', 'k2-with-cuda': True, 'k2-git-sha1': '8802dac96a4eabfc90e967dc8dd967f06b21985e', 'k2-git-date': 'Fri Feb 24 17:56:13 2023', 'lhotse-version': '1.15.0.dev+git.cbd7b2b.clean', 'torch-version': '1.12.0+cu113', 'torch-cuda-available': True, 'torch-cuda-version': '11.3', 'python-version': '3.8', 'icefall-git-branch': 'master', 'icefall-git-sha1': '3a1ce59-dirty', 'icefall-git-date': 'Tue Aug 29 14:09:48 2023', 'icefall-path': '/media/administrator/hdd/DeepSpeech/Sishir/icefall_aug29', 'k2-path': '/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/k2/__init__.py', 'lhotse-path': '/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/lhotse/__init__.py', 'hostname': 'administrator', 'IP address': '127.0.1.1'}, 'epoch': 30, 'iter': 0, 'avg': 15, 'use_averaged_model': True, 'exp_dir': PosixPath('pruned_transducer_stateless7/exp'), 'bpe_model': 'data/lang_bpe_500/bpe.model', 'lang_dir': PosixPath('data/lang_bpe_500'), 'decoding_method': 'modified_beam_search_lm_shallow_fusion', 'beam_size': 4, 'beam': 20.0, 'ngram_lm_scale': 0.01, 'max_contexts': 8, 'max_states': 64, 'context_size': 2, 'max_sym_per_frame': 1, 'num_paths': 200, 'nbest_scale': 0.5, 'use_shallow_fusion': True, 'lm_type': 'rnn', 'lm_scale': 0.3, 'tokens_ngram': 3, 'backoff_id': 500, 'num_encoder_layers': '2,4,3,2,4', 'feedforward_dims': '1024,1024,2048,2048,1024', 'nhead': '8,8,8,8,8', 'encoder_dims': '384,384,384,384,384', 'attention_dims': '192,192,192,192,192', 'encoder_unmasked_dims': '256,256,256,256,256', 'zipformer_downsampling_factors': '1,2,4,8,2', 'cnn_module_kernels': '31,31,31,31,31', 'decoder_dim': 512, 'joiner_dim': 512, 'full_libri': True, 'mini_libri': False, 'manifest_dir': PosixPath('data/fbank'), 'max_duration': 600, 'bucketing_sampler': True, 'num_buckets': 30, 'concatenate_cuts': False, 'duration_factor': 1.0, 'gap': 1.0, 'on_the_fly_feats': False, 'shuffle': True, 'drop_last': True, 'return_cuts': True, 'num_workers': 2, 'enable_spec_aug': True, 'spec_aug_time_warp_factor': 80, 'enable_musan': True, 'input_strategy': 'PrecomputedFeatures', 'lm_vocab_size': 500, 'lm_epoch': 7, 'lm_avg': 1, 'lm_exp_dir': None, 'rnn_lm_embedding_dim': 2048, 'rnn_lm_hidden_dim': 2048, 'rnn_lm_num_layers': 3, 'rnn_lm_tie_weights': True, 'transformer_lm_exp_dir': None, 'transformer_lm_dim_feedforward': 2048, 'transformer_lm_encoder_dim': 768, 'transformer_lm_embedding_dim': 768, 'transformer_lm_nhead': 8, 'transformer_lm_num_layers': 16, 'transformer_lm_tie_weights': True, 'res_dir': PosixPath('pruned_transducer_stateless7/exp/modified_beam_search_lm_shallow_fusion'), 'suffix': 'epoch-30-avg-15-modified_beam_search_lm_shallow_fusion-beam-size-4-rnnlm-lm-scale-0.3-use-averaged-model', 'blank_id': 0, 'unk_id': 2, 'vocab_size': 500}
2023-09-04 15:45:08,868 INFO [decode.py:795] About to create model
2023-09-04 15:45:09,173 INFO [zipformer.py:178] At encoder stack 4, which has downsampling_factor=2, we will combine the outputs of layers 1 and 3, with downsampling_factors=2 and 8.
2023-09-04 15:45:09,179 INFO [decode.py:862] Calculating the averaged model over epoch range from 15 (excluded) to 30
2023-09-04 15:45:17,399 INFO [model.py:75] Tying weights
2023-09-04 15:45:17,426 INFO [checkpoint.py:112] Loading checkpoint from None/epoch-7.pt
Traceback (most recent call last):
  File "pruned_transducer_stateless7/decode.py", line 962, in <module>
    main()
  File "/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "pruned_transducer_stateless7/decode.py", line 895, in main
    LM = LmScorer(
  File "/media/administrator/hdd/DeepSpeech/Sishir/icefall_aug29/icefall/lm_wrapper.py", line 45, in __init__
    self.lm = self.get_lm(lm_type, device, params)
  File "/media/administrator/hdd/DeepSpeech/Sishir/icefall_aug29/icefall/lm_wrapper.py", line 170, in get_lm
    load_checkpoint(
  File "/media/administrator/hdd/DeepSpeech/Sishir/icefall_aug29/icefall/checkpoint.py", line 113, in load_checkpoint
    checkpoint = torch.load(filename, map_location="cpu")
  File "/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/torch/serialization.py", line 699, in load
    with _open_file_like(f, 'rb') as opened_file:
  File "/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/torch/serialization.py", line 230, in _open_file_like
    return _open_file(name_or_buffer, mode)
  File "/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/torch/serialization.py", line 211, in __init__
    super(_open_file, self).__init__(open(name, mode))
FileNotFoundError: [Errno 2] No such file or directory: 'None/epoch-7.pt'

**Comment by csukuangfj** [comment]
Have you changed any code?
If yes, please show the output of
```
git diff
```

**Comment by csukuangfj** [comment]
> The argument --lm-exp-dir was in documentation.

Sorry, the documentation is not correct.

**Comment by bsshruthi22** [comment]
@csukuangfj   The --rnn-lm-tie-weights parameter is set to True by default in both training and decoding. The error I am getting is as below:
./pruned_transducer_stateless7/decode.py \
>     --epoch 30 \
>     --avg 15 \
>     --exp-dir ./pruned_transducer_stateless7/exp \
>     --max-duration 600 \
>     --decoding-method modified_beam_search_lm_shallow_fusion \
>     --bpe-model ./data/lang_bpe_500/bpe.model \
>     --beam-size 4 \
>     --lm-type rnn \
>     --lm-scale 0.3 \
>     --use-shallow-fusion True \
>     --lm-exp-dir ./pruned_transducer_stateless7/rnn_lm/exp \
>     --rnn-lm-num-layers 3 \
>     --rnn-lm-tie-weights 1
2023-09-07 17:11:29,698 INFO [decode.py:778] Decoding started
2023-09-07 17:11:29,698 INFO [decode.py:784] Device: cuda:0
2023-09-07 17:11:29,699 INFO [decode.py:794] {'frame_shift_ms': 10.0, 'allowed_excess_duration_ratio': 0.1, 'best_train_loss': inf, 'best_valid_loss': inf, 'best_train_epoch': -1, 'best_valid_epoch': -1, 'batch_idx_train': 0, 'log_interval': 50, 'reset_interval': 200, 'valid_interval': 3000, 'feature_dim': 80, 'subsampling_factor': 4, 'warm_step': 2000, 'env_info': {'k2-version': '1.23.4', 'k2-build-type': 'Release', 'k2-with-cuda': True, 'k2-git-sha1': '8802dac96a4eabfc90e967dc8dd967f06b21985e', 'k2-git-date': 'Fri Feb 24 17:56:13 2023', 'lhotse-version': '1.15.0.dev+git.cbd7b2b.clean', 'torch-version': '1.12.0+cu113', 'torch-cuda-available': True, 'torch-cuda-version': '11.3', 'python-version': '3.8', 'icefall-git-branch': 'master', 'icefall-git-sha1': '3a1ce59-dirty', 'icefall-git-date': 'Tue Aug 29 14:09:48 2023', 'icefall-path': '/media/administrator/hdd/DeepSpeech/Sishir/icefall_aug29', 'k2-path': '/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/k2/__init__.py', 'lhotse-path': '/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/lhotse/__init__.py', 'hostname': 'administrator', 'IP address': '127.0.1.1'}, 'epoch': 30, 'iter': 0, 'avg': 15, 'use_averaged_model': True, 'exp_dir': PosixPath('pruned_transducer_stateless7/exp'), 'bpe_model': './data/lang_bpe_500/bpe.model', 'lang_dir': PosixPath('data/lang_bpe_500'), 'decoding_method': 'modified_beam_search_lm_shallow_fusion', 'beam_size': 4, 'beam': 20.0, 'ngram_lm_scale': 0.01, 'max_contexts': 8, 'max_states': 64, 'context_size': 2, 'max_sym_per_frame': 1, 'num_paths': 200, 'nbest_scale': 0.5, 'use_shallow_fusion': True, 'lm_type': 'rnn', 'lm_scale': 0.3, 'tokens_ngram': 3, 'backoff_id': 500, 'num_encoder_layers': '2,4,3,2,4', 'feedforward_dims': '1024,1024,2048,2048,1024', 'nhead': '8,8,8,8,8', 'encoder_dims': '384,384,384,384,384', 'attention_dims': '192,192,192,192,192', 'encoder_unmasked_dims': '256,256,256,256,256', 'zipformer_downsampling_factors': '1,2,4,8,2', 'cnn_module_kernels': '31,31,31,31,31', 'decoder_dim': 512, 'joiner_dim': 512, 'full_libri': True, 'mini_libri': False, 'manifest_dir': PosixPath('data/fbank'), 'max_duration': 600, 'bucketing_sampler': True, 'num_buckets': 30, 'concatenate_cuts': False, 'duration_factor': 1.0, 'gap': 1.0, 'on_the_fly_feats': False, 'shuffle': True, 'drop_last': True, 'return_cuts': True, 'num_workers': 2, 'enable_spec_aug': True, 'spec_aug_time_warp_factor': 80, 'enable_musan': True, 'input_strategy': 'PrecomputedFeatures', 'lm_vocab_size': 500, 'lm_epoch': 7, 'lm_avg': 1, 'lm_exp_dir': './pruned_transducer_stateless7/rnn_lm/exp', 'rnn_lm_embedding_dim': 2048, 'rnn_lm_hidden_dim': 2048, 'rnn_lm_num_layers': 3, 'rnn_lm_tie_weights': True, 'transformer_lm_exp_dir': None, 'transformer_lm_dim_feedforward': 2048, 'transformer_lm_encoder_dim': 768, 'transformer_lm_embedding_dim': 768, 'transformer_lm_nhead': 8, 'transformer_lm_num_layers': 16, 'transformer_lm_tie_weights': True, 'res_dir': PosixPath('pruned_transducer_stateless7/exp/modified_beam_search_lm_shallow_fusion'), 'suffix': 'epoch-30-avg-15-modified_beam_search_lm_shallow_fusion-beam-size-4-rnnlm-lm-scale-0.3-use-averaged-model', 'blank_id': 0, 'unk_id': 2, 'vocab_size': 500}
2023-09-07 17:11:29,700 INFO [decode.py:796] About to create model
2023-09-07 17:11:30,001 INFO [zipformer.py:178] At encoder stack 4, which has downsampling_factor=2, we will combine the outputs of layers 1 and 3, with downsampling_factors=2 and 8.
2023-09-07 17:11:30,007 INFO [decode.py:863] Calculating the averaged model over epoch range from 15 (excluded) to 30
2023-09-07 17:11:32,845 INFO [model.py:75] Tying weights
2023-09-07 17:11:32,845 INFO [checkpoint.py:112] Loading checkpoint from ./pruned_transducer_stateless7/rnn_lm/exp/epoch-7.pt
Traceback (most recent call last):
  File "./pruned_transducer_stateless7/decode.py", line 964, in <module>
    main()
  File "/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "./pruned_transducer_stateless7/decode.py", line 896, in main
    LM = LmScorer(
  File "/media/administrator/hdd/DeepSpeech/Sishir/icefall_aug29/icefall/lm_wrapper.py", line 45, in __init__
    self.lm = self.get_lm(lm_type, device, params)
  File "/media/administrator/hdd/DeepSpeech/Sishir/icefall_aug29/icefall/lm_wrapper.py", line 170, in get_lm
    load_checkpoint(
  File "/media/administrator/hdd/DeepSpeech/Sishir/icefall_aug29/icefall/checkpoint.py", line 126, in load_checkpoint
    model.load_state_dict(checkpoint["model"], strict=strict)
  File "/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1604, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for RnnLmModel:
	size mismatch for input_embedding.weight: copying a param with shape torch.Size([500, 800]) from checkpoint, the shape in current model is torch.Size([500, 2048]).
	size mismatch for rnn.weight_ih_l0: copying a param with shape torch.Size([800, 800]) from checkpoint, the shape in current model is torch.Size([8192, 2048]).
	size mismatch for rnn.weight_hh_l0: copying a param with shape torch.Size([800, 200]) from checkpoint, the shape in current model is torch.Size([8192, 2048]).
	size mismatch for rnn.bias_ih_l0: copying a param with shape torch.Size([800]) from checkpoint, the shape in current model is torch.Size([8192]).
	size mismatch for rnn.bias_hh_l0: copying a param with shape torch.Size([800]) from checkpoint, the shape in current model is torch.Size([8192]).
	size mismatch for rnn.weight_ih_l1: copying a param with shape torch.Size([800, 200]) from checkpoint, the shape in current model is torch.Size([8192, 2048]).
	size mismatch for rnn.weight_hh_l1: copying a param with shape torch.Size([800, 200]) from checkpoint, the shape in current model is torch.Size([8192, 2048]).
	size mismatch for rnn.bias_ih_l1: copying a param with shape torch.Size([800]) from checkpoint, the shape in current model is torch.Size([8192]).
	size mismatch for rnn.bias_hh_l1: copying a param with shape torch.Size([800]) from checkpoint, the shape in current model is torch.Size([8192]).
	size mismatch for output_linear.weight: copying a param with shape torch.Size([500, 200]) from checkpoint, the shape in current model is torch.Size([500, 2048]).

**Comment by csukuangfj** [comment]
> @csukuangfj   The --rnn-lm-tie-weights parameter is set to True by default in both training and decoding. The error I am getting is as below:
> 
> ./pruned_transducer_stateless7/decode.py \
> 
> >     --epoch 30 \
> 
> >     --avg 15 \
> 
> >     --exp-dir ./pruned_transducer_stateless7/exp \
> 
> >     --max-duration 600 \
> 
> >     --decoding-method modified_beam_search_lm_shallow_fusion \
> 
> >     --bpe-model ./data/lang_bpe_500/bpe.model \
> 
> >     --beam-size 4 \
> 
> >     --lm-type rnn \
> 
> >     --lm-scale 0.3 \
> 
> >     --use-shallow-fusion True \
> 
> >     --lm-exp-dir ./pruned_transducer_stateless7/rnn_lm/exp \
> 
> >     --rnn-lm-num-layers 3 \
> 
> >     --rnn-lm-tie-weights 1
> 
> 2023-09-07 17:11:29,698 INFO [decode.py:778] Decoding started
> 
> 2023-09-07 17:11:29,698 INFO [decode.py:784] Device: cuda:0
> 
> 2023-09-07 17:11:29,699 INFO [decode.py:794] {'frame_shift_ms': 10.0, 'allowed_excess_duration_ratio': 0.1, 'best_train_loss': inf, 'best_valid_loss': inf, 'best_train_epoch': -1, 'best_valid_epoch': -1, 'batch_idx_train': 0, 'log_interval': 50, 'reset_interval': 200, 'valid_interval': 3000, 'feature_dim': 80, 'subsampling_factor': 4, 'warm_step': 2000, 'env_info': {'k2-version': '1.23.4', 'k2-build-type': 'Release', 'k2-with-cuda': True, 'k2-git-sha1': '8802dac96a4eabfc90e967dc8dd967f06b21985e', 'k2-git-date': 'Fri Feb 24 17:56:13 2023', 'lhotse-version': '1.15.0.dev+git.cbd7b2b.clean', 'torch-version': '1.12.0+cu113', 'torch-cuda-available': True, 'torch-cuda-version': '11.3', 'python-version': '3.8', 'icefall-git-branch': 'master', 'icefall-git-sha1': '3a1ce59-dirty', 'icefall-git-date': 'Tue Aug 29 14:09:48 2023', 'icefall-path': '/media/administrator/hdd/DeepSpeech/Sishir/icefall_aug29', 'k2-path': '/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/k2/__init__.py', 'lhotse-path': '/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/lhotse/__init__.py', 'hostname': 'administrator', 'IP address': '127.0.1.1'}, 'epoch': 30, 'iter': 0, 'avg': 15, 'use_averaged_model': True, 'exp_dir': PosixPath('pruned_transducer_stateless7/exp'), 'bpe_model': './data/lang_bpe_500/bpe.model', 'lang_dir': PosixPath('data/lang_bpe_500'), 'decoding_method': 'modified_beam_search_lm_shallow_fusion', 'beam_size': 4, 'beam': 20.0, 'ngram_lm_scale': 0.01, 'max_contexts': 8, 'max_states': 64, 'context_size': 2, 'max_sym_per_frame': 1, 'num_paths': 200, 'nbest_scale': 0.5, 'use_shallow_fusion': True, 'lm_type': 'rnn', 'lm_scale': 0.3, 'tokens_ngram': 3, 'backoff_id': 500, 'num_encoder_layers': '2,4,3,2,4', 'feedforward_dims': '1024,1024,2048,2048,1024', 'nhead': '8,8,8,8,8', 'encoder_dims': '384,384,384,384,384', 'attention_dims': '192,192,192,192,192', 'encoder_unmasked_dims': '256,256,256,256,256', 'zipformer_downsampling_factors': '1,2,4,8,2', 'cnn_module_kernels': '31,31,31,31,31', 'decoder_dim': 512, 'joiner_dim': 512, 'full_libri': True, 'mini_libri': False, 'manifest_dir': PosixPath('data/fbank'), 'max_duration': 600, 'bucketing_sampler': True, 'num_buckets': 30, 'concatenate_cuts': False, 'duration_factor': 1.0, 'gap': 1.0, 'on_the_fly_feats': False, 'shuffle': True, 'drop_last': True, 'return_cuts': True, 'num_workers': 2, 'enable_spec_aug': True, 'spec_aug_time_warp_factor': 80, 'enable_musan': True, 'input_strategy': 'PrecomputedFeatures', 'lm_vocab_size': 500, 'lm_epoch': 7, 'lm_avg': 1, 'lm_exp_dir': './pruned_transducer_stateless7/rnn_lm/exp', 'rnn_lm_embedding_dim': 2048, 'rnn_lm_hidden_dim': 2048, 'rnn_lm_num_layers': 3, 'rnn_lm_tie_weights': True, 'transformer_lm_exp_dir': None, 'transformer_lm_dim_feedforward': 2048, 'transformer_lm_encoder_dim': 768, 'transformer_lm_embedding_dim': 768, 'transformer_lm_nhead': 8, 'transformer_lm_num_layers': 16, 'transformer_lm_tie_weights': True, 'res_dir': PosixPath('pruned_transducer_stateless7/exp/modified_beam_search_lm_shallow_fusion'), 'suffix': 'epoch-30-avg-15-modified_beam_search_lm_shallow_fusion-beam-size-4-rnnlm-lm-scale-0.3-use-averaged-model', 'blank_id': 0, 'unk_id': 2, 'vocab_size': 500}
> 
> 2023-09-07 17:11:29,700 INFO [decode.py:796] About to create model
> 
> 2023-09-07 17:11:30,001 INFO [zipformer.py:178] At encoder stack 4, which has downsampling_factor=2, we will combine the outputs of layers 1 and 3, with downsampling_factors=2 and 8.
> 
> 2023-09-07 17:11:30,007 INFO [decode.py:863] Calculating the averaged model over epoch range from 15 (excluded) to 30
> 
> 2023-09-07 17:11:32,845 INFO [model.py:75] Tying weights
> 
> 2023-09-07 17:11:32,845 INFO [checkpoint.py:112] Loading checkpoint from ./pruned_transducer_stateless7/rnn_lm/exp/epoch-7.pt
> 
> Traceback (most recent call last):
> 
>   File "./pruned_transducer_stateless7/decode.py", line 964, in <module>
> 
>     main()
> 
>   File "/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
> 
>     return func(*args, **kwargs)
> 
>   File "./pruned_transducer_stateless7/decode.py", line 896, in main
> 
>     LM = LmScorer(
> 
>   File "/media/administrator/hdd/DeepSpeech/Sishir/icefall_aug29/icefall/lm_wrapper.py", line 45, in __init__
> 
>     self.lm = self.get_lm(lm_type, device, params)
> 
>   File "/media/administrator/hdd/DeepSpeech/Sishir/icefall_aug29/icefall/lm_wrapper.py", line 170, in get_lm
> 
>     load_checkpoint(
> 
>   File "/media/administrator/hdd/DeepSpeech/Sishir/icefall_aug29/icefall/checkpoint.py", line 126, in load_checkpoint
> 
>     model.load_state_dict(checkpoint["model"], strict=strict)
> 
>   File "/home/administrator/anaconda3/envs/k2/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1604, in load_state_dict
> 
>     raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
> 
> RuntimeError: Error(s) in loading state_dict for RnnLmModel:
> 
> 	size mismatch for input_embedding.weight: copying a param with shape torch.Size([500, 800]) from checkpoint, the shape in current model is torch.Size([500, 2048]).
> 
> 	size mismatch for rnn.weight_ih_l0: copying a param with shape torch.Size([800, 800]) from checkpoint, the shape in current model is torch.Size([8192, 2048]).
> 
> 	size mismatch for rnn.weight_hh_l0: copying a param with shape torch.Size([800, 200]) from checkpoint, the shape in current model is torch.Size([8192, 2048]).
> 
> 	size mismatch for rnn.bias_ih_l0: copying a param with shape torch.Size([800]) from checkpoint, the shape in current model is torch.Size([8192]).
> 
> 	size mismatch for rnn.bias_hh_l0: copying a param with shape torch.Size([800]) from checkpoint, the shape in current model is torch.Size([8192]).
> 
> 	size mismatch for rnn.weight_ih_l1: copying a param with shape torch.Size([800, 200]) from checkpoint, the shape in current model is torch.Size([8192, 2048]).
> 
> 	size mismatch for rnn.weight_hh_l1: copying a param with shape torch.Size([800, 200]) from checkpoint, the shape in current model is torch.Size([8192, 2048]).
> 
> 	size mismatch for rnn.bias_ih_l1: copying a param with shape torch.Size([800]) from checkpoint, the shape in current model is torch.Size([8192]).
> 
> 	size mismatch for rnn.bias_hh_l1: copying a param with shape torch.Size([800]) from checkpoint, the shape in current model is torch.Size([8192]).
> 
> 	size mismatch for output_linear.weight: copying a param with shape torch.Size([500, 200]) from checkpoint, the shape in current model is torch.Size([500, 2048]).
> 
> 

Please post the command you used to train the LM.

**Comment by marcoyang1998** [comment]
@bsshruthi22  You are not passing the arguments about the rnnlm when calling `decode.py`.

According to the training command you posted at the beginning, you should specify the following arguments:
```
--epoch 30 \
--avg 15 \
--exp-dir ./pruned_transducer_stateless7/exp \
--max-duration 600 \
--decoding-method modified_beam_search_lm_shallow_fusion \
--bpe-model ./data/lang_bpe_500/bpe.model \
--beam-size 4 \
--lm-type rnn \
--lm-scale 0.3 \
--use-shallow-fusion True \
--lm-exp-dir ./pruned_transducer_stateless7/rnn_lm/exp \
--rnn-lm-num-layers 2  \
--rnn-lm-tie-weights 0  \
--rnn-lm-embedding-dim 800 \
--rnn-lm-hidden-dim 200 \
--rnn-lm-num-layers 2
```

You may find the possible arguments here: https://github.com/k2-fsa/icefall/blob/49a4b672884213809cc04df2caab6c37cee92c22/icefall/lm_wrapper.py#L50
